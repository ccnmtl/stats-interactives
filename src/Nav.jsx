import React from 'react';
import {Link} from 'react-router-dom';

export const Nav = () => (
    <nav className="navbar navbar-expand-lg navbar-dark">
        <Link className="navbar-brand" to="/">
            <span className="stats">Stats</span>
            Interactives
            <span className="beta"> Beta</span>
        </Link>
        <div className="navbar-nav-scroll">
            <ul className="navbar-nav bd-navbar-nav flex-row">
                <li className="nav-item">
                    <Link to="/" className="nav-link">
                        &larr; Back to the Collection
                    </Link>
                </li>
            </ul>
        </div>
        <button
            className="navbar-toggler"
            type="button"
            data-toggle="collapse"
            data-target="#navbarSupportedContent"
            aria-controls="navbarSupportedContent"
            aria-expanded="false" aria-label="Toggle navigation">
            <span className="navbar-toggler-icon"></span> </button>
        <div className="collapse navbar-collapse" id="navbarSupportedContent">
        </div>
    </nav>
);
