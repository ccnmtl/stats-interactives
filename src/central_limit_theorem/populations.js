// This contains methods for generating the perfect, theoretical histogram
// for a given population.

function GetZPercent(z) {
    if (z < -6.5) {
        return 0.0;
    } else if (z > 6.5) {
        return 1.0;
    }
    let factK = 1;
    let sum = 0;
    let term = 1;
    let k = 0;
    let loopStop = Math.exp(-23);
    while(Math.abs(term) > loopStop) {
        /* eslint-disable-next-line */
        term = .3989422804 * Math.pow(-1, k) * Math.pow(z, k) / (2 * k + 1) / Math.pow(2, k) * Math.pow(z, k + 1) / factK;
        sum += term;
        k++;
        factK *= k;
    }
    sum += 0.5;
    return sum;
}

export const normalBarHeight = (intervals, mu, sigma) => {
    /* @intervals = An array of the values of the intervals to fill
     * @mu = mean of the normal distribution
     * @sigma = standard deviation of the normal ditribution
     *
     * returns an array of values [0, 1] that illustrate a theoritically
     * perfect normal distribution
     */
    let mean = mu || 0.0;
    let sd = sigma || 1.0;

    const SCALING_FACTOR = 0.341344746067521438;
    return [...Array(intervals.length - 1)].map((val, idx) => {
        let barMinZ = (intervals[idx] - mean) / sd;
        let barMaxZ = (intervals[idx + 1] - mean) / sd;
        return (GetZPercent(barMaxZ) - GetZPercent(barMinZ)) / SCALING_FACTOR;
    });
};

export const exponentialBarHeight = (intervals, mu, sigma) => {
    /* @intervals = An array of the values of the intervals to fill
     * @mu = mean of the normal distribution
     * @sigma = standard deviation of the normal ditribution
     *
     * returns an array of values [0, 1] that illustrate a theoritically
     * perfect exponential distribution
     */
    let mean = mu || 0.0;
    let sd = sigma || 1.0;

    const SCALING_FACTOR = 0.6321205588285577;
    intervals = intervals.map((val) => {
        return val < mean - sd ? mean - sd : val;
    });
    return [...Array(intervals.length - 1)].map((val, idx) => {
        return (1 - Math.exp(-1 / sd * (intervals[idx + 1] - mean + sd)) -
                (1 - Math.exp(-1 / sd * (intervals[idx] - mean + sd)))) /
                    SCALING_FACTOR;
    });
};
