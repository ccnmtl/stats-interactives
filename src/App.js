import { CentralLimitGraph } from './CentralLimitGraph'
import { BrowserRouter as Router, Route, Link } from 'react-router-dom';

export const App = () => (
    <Router>
        <div>
            <h1>Graphs</h1>
            <div className="container">
                <nav>
                    <ul>
                        <li>
                            <Link to="/">Home</Link>
                        </li>
                        <li>
                            <Link to="/central-limit-theorem">Central Limit Theorem</Link>
                        </li>
                    </ul>
                </nav>
            </div>
            <div className="container">
                <Route exact path="/" component={Home} />
                <Route path="/central-limit-theorem" component={CentralLimitGraph} />
            </div>
        </div>
    </Router>
)

const Home = () => (
    <h1>Home!</h1>
)
