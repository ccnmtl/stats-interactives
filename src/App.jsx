/*eslint max-len: ["error", { "ignoreStrings": true }]*/
import React from 'react';
import PropTypes from 'prop-types';
import { CentralLimitGraph } from './CentralLimitGraph';
import { BrowserRouter as Router, Route, Link } from 'react-router-dom';

export const App = () => (
    <Router>
        <main role="main">
            <Route exact path="/" component={Home} />
            <Route path="/central-limit-theorem"
                component={CentralLimitGraph} />
        </main>
    </Router>
);

const PreviewData = [
    {
        'slug': '/central-limit-theorem',
        'title': 'Central Limit Theorem',
        'text': 'Donec id elit non mi porta gravida at eget metus. Fusce dapibus, tellus ac cursus commodo, tortor mauris condimentum nibh, ut fermentum massa justo sit amet risus. Etiam porta sem malesuada magna mollis euismod. Donec sed odio dui.'
    },
    {
        'slug': '/ols-regression',
        'title': 'OLS Regression',
        'text': 'Donec id elit non mi porta gravida at eget metus. Fusce dapibus, tellus ac cursus commodo, tortor mauris condimentum nibh, ut fermentum massa justo sit amet risus. Etiam porta sem malesuada magna mollis euismod. Donec sed odio dui.'
    },
];

const PreviewBlock = (props) => (
    <div key={props.slug} className="col-md-4">
        <h2>
            <Link to={props.slug}>{ props.title }</Link>
        </h2>
        <p>{props.text}</p>
    </div>
);

const Preview = () => (
    PreviewData.map(PreviewBlock)
);

const Home = () => (
    <>
    <div className="jumbotron">
    
        <div className="container">
        
            <h1 className="display-3">
                Theorem7
            </h1>
            
            <p className="lead">
                Seven Interactive Simulations of Statistics Theorems for Deeper Understanding
            </p>

            <p>
                Specifically designed for Doru Cojoc&#39;s Quantitative
                Analysis I course at SIPA, these simulations are intended as a
                test of effectiveness at improving learning outcomes. This
                project seeks to address this by turning them into web
                interactives, openly available. These interactives will be
                provided to students for their own study and experimentation,
                which can be used here or embedded/linked elsewhere.
            </p>
            
            <p>
                <a className="btn btn-primary btn-lg" href="https://sipa.columbia.edu/faculty-research/faculty-directory/doru-c-cojoc" role="button">Contact Dr. Cojoc</a>
            </p>
            
        </div>
        
    </div>
    
    <div className="container">
    
        <div className="row">
        
            <Preview />
            
        </div>
        
    </div>
    </>
);

PreviewBlock.propTypes = {
    slug: PropTypes.string,
    title: PropTypes.string,
    text: PropTypes.string
};
export {PreviewBlock, Preview};
