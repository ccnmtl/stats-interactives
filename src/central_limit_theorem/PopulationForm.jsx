import React from 'react';
import PropTypes from 'prop-types';
import { forceNumber } from '../utils';
import { LabeledSlider } from '../LabeledSlider';
import { DISTRIBUTION_TYPE } from './CentralLimitGraph';

export const PopulationForm  = (
    {seed, populationSize, mean, stdDev, distType,
        sampleSize, handleChange, handleGeneratePopulation, showPopBtn}) => {
    const handleFormChange = (e) => {
        let numericFields = ['populationSize', 'mean', 'stdDev'];
        if (numericFields.includes(e.target.id)) {
            handleChange(e.target.id, forceNumber(e.target.value));
        } else {
            handleChange(e.target.id, e.target.value);
        }
    };
    const handleRheostatChange = (sliderState) => {
        handleChange('mean', sliderState.values[0]);
    };
    const handleGenPop = (e) => {
        e.preventDefault();
        handleGeneratePopulation();
    };
    return (
        <>
        <form onSubmit={handleGenPop}>
            <fieldset>
                <legend>
                Set the parameters.
                </legend>
                <div className="form-row">
                    <div className="form-group col-md-5">
                        <label htmlFor="seed" className="float-right">
                            Seed: </label>
                    </div>
                    <div className="form-group col-md-7">
                        <input type="text"
                            id="seed"
                            value={seed}
                            onChange={handleFormChange}/>
                    </div>
                </div>
                <div className="form-row">
                    <div className="form-group col-md-5">
                        <label htmlFor="populationSize"
                            className="float-right">
                            Population Size: </label>
                    </div>
                    <div className="form-group col-md-7">
                        <input type="number"
                            id="populationSize"
                            disabled={seed ? false : true}
                            value={populationSize}
                            onChange={handleFormChange}/>
                    </div>
                </div>
                <div className="form-row">
                    <div className="form-group col-md-5">
                        <label htmlFor="distType"
                            className="float-right">
                            Distribution Type: </label>
                    </div>
                    <div className="form-group col-md-7">
                        <select id="distType"
                            onChange={handleFormChange}
                            disabled={seed ? false : true}
                            value={distType}>
                            { DISTRIBUTION_TYPE.map(
                                (e) => (<option key={e.value} value={e.value}>
                                    {e.display}</option>)) }
                        </select>
                    </div>
                </div>
                <div className="form-row">
                    <div className="form-group col-md-12">
                        <label htmlFor="mean"
                            className="float-right">Mean:</label>
                    </div>
                    <div className="form-group col-md-12">
                        {/*
                        <input type="range"
                            disabled={seed ? false : true}
                            id="mean"
                            min="-10"
                            max="10"
                            value={mean}
                            onChange={handleFormChange}/> */}
                        <LabeledSlider
                            disabled={seed ? false : true}
                            min={-10}
                            max={10}
                            values={[mean]}
                            snap
                            pitPoints={[-10, -9, -8, -7, -6, -5, -4, -3, -2,
                                -1, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]}
                            snapPoints={[-10, -9, -8, -7, -6, -5, -4, -3, -2,
                                -1, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]}
                            onChange={handleRheostatChange}/>
                    </div>
                </div>
                <div className="form-row">
                    <div className="form-group col-md-5">
                        <label htmlFor="stdDev"
                            className="float-right">StdDev: </label>
                    </div>
                    <div className="form-group col-md-7">
                        <input type="range"
                            disabled={seed ? false : true}
                            id="stdDev"
                            min="-6"
                            max="6"
                            value={stdDev}
                            onChange={handleFormChange}/>
                    </div>
                </div>
                <div className="form-row">
                    <div className="form-group offset-md-5 col-md-7">
                        <input className="btn btn-primary"
                            disabled={seed ? false : true}
                            id="generate-population"
                            type="submit"
                            value="Generate Population"/>
                    </div>
                </div>
            </fieldset>
        </form>
        </>
    );
};

PopulationForm.propTypes = {
    seed: PropTypes.string,
    populationSize: PropTypes.number,
    mean: PropTypes.number,
    stdDev: PropTypes.number,
    distType: PropTypes.string,
    handleChange: PropTypes.func,
    showPopBtn: PropTypes.bool,
    sampleSize: PropTypes.number,
    handleGeneratePopulation: PropTypes.func
};
