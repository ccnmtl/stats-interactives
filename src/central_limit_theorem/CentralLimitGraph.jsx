import React, { Component } from 'react';
import * as math from 'mathjs';
import {
    createHistogramArray, getHistogramMaxima,
    createScatterPlotHistogram } from '../utils.js';
import { Nav } from '../Nav.jsx';
import { PopulationGraph } from './PopulationGraph';
import { SampleMeansGraph } from './SampleMeansGraph';
import { PopulationForm } from './PopulationForm';
import { SampleForm } from './SampleForm';
import { SampleRangeSlider } from './SampleRangeSlider';
import {CopyToClipboard} from 'react-copy-to-clipboard';

var seedrandom = require('seedrandom');
var jStat = require('jStat').jStat;

export const MIN_BIN = -18;
export const MAX_BIN = 18;
export const NO_OF_BINS = MAX_BIN - MIN_BIN;

export const DISTRIBUTION_TYPE = [
    {
        value: 'normal',
        display: 'Normal',
    },
    {
        value: 'skew_left',
        display: 'Skew Left',
    },
    {
        value: 'skew_right',
        display: 'Skew Right',
    },
    {
        value: 'bimodal',
        display: 'Bimodal'
    },
    {
        value: 'uniform',
        display: 'Uniform',
    },
];

export class CentralLimitGraph extends Component {
    constructor(props) {
        super(props);

        this.handleChange = this.handleChange.bind(this);
        this.generatePopulation = this.generatePopulation.bind(this);
        this.handleGeneratePopulation = this.handleGeneratePopulation
            .bind(this);
        this.handleResetPopulation = this.handleResetPopulation.bind(this);
        this.runSample = this.runSample.bind(this);
        this.handleSampleMeansIdx = this.handleSampleMeansIdx.bind(this);
        this.handleObservationIdx = this.handleObservationIdx.bind(this);
        this.handleResetSamples = this.handleResetSamples.bind(this);
        this.handleResetSimulation = this.handleResetSimulation.bind(this);

        let params = new URLSearchParams(location.search);
        let seed = '';
        if (!params.has('seed')) {
            // generate a seed
            seed = String(Date.now());
            params.set('seed', seed);
            window.history.replaceState(null, '', '?' + params.toString());
        } else {
            seed = String(params.get('seed'));
        }

        let distType = '';
        if (!params.has('distType')) {
            distType = 'skew_left';
            params.set('distType', 'skew_left');
            window.history.replaceState(null, '', '?' + params.toString());
        } else {
            // Validate that a dist type is on the list, else
            // set it to 'skew_left'
            let dt = params.get('distType');
            let found = DISTRIBUTION_TYPE.find(
                (e) => {return dt === e.value;});
            distType = found ? dt : 'skew_left';
            if (!found) {
                params.set('distType', 'skew_left');
                window.history.replaceState(null, '', '?' + params.toString());
            }
        }

        const populationSize = 100000;
        const mean = 0;
        const stdDev = 1;

        const defaultSampleSize = 50;
        const defaultNumberOfSamples = 1000;

        this.state = {
            seed: seed,
            populationSize: populationSize,
            population: null,
            populationGraphData: null,
            mean: mean,
            stdDev: stdDev,
            distType: distType,
            // sampleSize is the number of observations within each sample
            sampleSize: defaultSampleSize,
            // number of samples is the overall samples taken of a population
            numberOfSamples: defaultNumberOfSamples,
            samples: null,
            samplesIdx: 1,
            samplesGraphData: null,
            sampleMeans: null,
            sampleMeansIdx: null,
            sampleMeansGraphData: null,
            domain: [-6, 6],
            sampleMeansDomain: null,
            sampleMeansRange: [0, 1],
            observationIdx: null,
            observationData: null,
        };
    }
    handleChange(key, value) {
        this.setState({
            [key]: value
        });
    }
    handleGeneratePopulation() {
        let population = this.generatePopulation();
        let populationGraphData = createHistogramArray(
            population,
            NO_OF_BINS,
            MIN_BIN,
            MAX_BIN
        );
        this.setState({
            population: population,
            populationGraphData: populationGraphData,
            domain: [Math.min(...population), Math.max(...population)]
        });
    }
    handleResetPopulation() {
        this.setState({
            population: null,
            populationGraphData: null,
            domain: null
        });
    }
    generatePopulation() {
        this.handleResetSamples();
        let size = this.state.populationSize;
        let mean = this.state.mean;
        let stdDev = this.state.stdDev;
        let distType = this.state.distType;
        let seed = this.state.seed;
        let rate = 1 / stdDev;

        // Reset the global Math.random everytime this is called
        seedrandom(seed, {global: true});

        switch (distType) {
        case 'normal':
            // This method munges the normal distribution such that every
            // value generated that is less than the mean is reflected about
            // the mean to produce a symetrical distribution.

            // Generate the first half of a normal distribution
            return [...Array(Math.floor(size / 2))].map((e) => {
                let s = null;
                do {
                    s = math.round(jStat.normal.sample(mean, stdDev), 1);
                } while (s > mean);
                return s;
            }).reduce((acc, e) => {
                // Then push the element on the accumulated array
                acc.push(e);
                if (e < mean) {
                    // If e is less than the mean, then find its reflection
                    // and push that to the array.
                    let diff = math.round(mean - e, 1);
                    acc.push(math.round(mean + diff, 1));
                }
                return acc;
            }, []);

        case 'skew_left':
            return [...Array(size)].map((e) => {
                return math.round(
                    jStat.exponential.sample(rate) - stdDev + mean, 1);
            });

        case 'skew_right':
            return [...Array(size)].map((e) => {
                return math.round(
                    (jStat.exponential.sample(rate) * -1) - stdDev + mean, 1);
            });

        case 'bimodal':
            return [...Array(size)].map((e, i) => {
                if (i % 2) {
                    return math.round(jStat.normal.sample(-3, stdDev), 1);
                } else {
                    return math.round(jStat.normal.sample(3, stdDev), 1);
                }
            });

        case 'uniform':
            return [...Array(size)].map((e) => {
                // Oversample the uniform distribution and then 'trim' off
                // extra values. This ensures that rounding doesn't taper the
                // distributions at the ends.
                let s = null;
                do {
                    s = math.round(jStat.uniform.sample(-1.1, 1.1), 1);
                } while (Math.abs(s) > 1);
                return s;
            });

        default:
            // return a normal distribution
            return [...Array(size)].map((e) => {
                return math.round(jStat.normal.sample(mean, stdDev), 1);
            });
        }

    }
    runSample() {
        // Use the base64 encoding of the seed as a simple hash
        let samplingSeed = window.btoa(this.state.seed);
        let ng = seedrandom(samplingSeed);

        // take samples from population values
        // - one sample consists of N sampleSize, which are then averaged
        // - N numberOfSamples are taken
        // push these to sampleSet
        // get the histogram and set state, render a line graph
        let samples = [...Array(this.state.numberOfSamples)].map((e) => {
            return [...Array(this.state.sampleSize)].map((e) => {
                let idx = Math.floor(ng() * this.state.population.length);
                return this.state.population[idx];
            });
        });

        let sampleMeans = samples.reduce((acc, e) => {
            acc.push(
                math.round(jStat.mean(e), 1)
            );
            return acc;
        }, []);

        let samplesGraphData = createScatterPlotHistogram(
            samples[0], NO_OF_BINS, MIN_BIN, MAX_BIN);

        let samplesMaxFrequency = 0;
        samples.map((e) => {
            let max = getHistogramMaxima(createScatterPlotHistogram(
                e, NO_OF_BINS, MIN_BIN, MAX_BIN));
            if (max > samplesMaxFrequency) {
                samplesMaxFrequency = max;
            }
        });

        this.setState({
            samples: samples,
            sampleMeans: sampleMeans,
            sampleMeansDomain: [
                Math.min(...sampleMeans),
                Math.max(...sampleMeans)
            ],
            sampleMeansRange: [
                0,
                getHistogramMaxima(createHistogramArray(
                    sampleMeans,
                    NO_OF_BINS,
                    MIN_BIN,
                    MAX_BIN
                ))
            ],
            sampleMeansIdx: 1,
            samplesGraphData: samplesGraphData,
            sampleMeansGraphData: sampleMeans.slice(0, 1),
            samplesMax: samplesMaxFrequency,
            observationIdx: 1,
            observationData: [samplesGraphData[0]],
        });
    }
    handleSampleMeansIdx(idx) {
        let currentSampleMeans = this.state.sampleMeans.slice(0, idx);
        let currentSampleMeansData = createHistogramArray(
            currentSampleMeans,
            NO_OF_BINS,
            MIN_BIN,
            MAX_BIN);
        let samplesGraphData = createScatterPlotHistogram(
            this.state.samples[idx - 1],
            13,
            this.state.domain[0],
            this.state.domain[1]);

        this.setState({
            sampleMeansIdx: idx,
            sampleMeansGraphData: currentSampleMeansData,
            samplesGraphData: samplesGraphData,
            observationData: [samplesGraphData[this.state.observationIdx -1]],
        });
    }
    handleObservationIdx(idx) {
        this.setState({
            observationIdx: idx,
            observationData: [this.state.samplesGraphData[idx - 1]]
        });
    }
    handleResetSamples() {
        this.setState({
            samples: null,
            sampleMeans: null,
            sampleMeansRange: null,
            sampleMeansIdx: null,
            samplesGraphData: null,
            sampleMeansGraphData: null,
            observationIdx: null,
            observationData: null,
        });
    }
    handleResetSimulation() {
        this.handleResetSamples();
        this.handleResetPopulation();
    }
    componentDidUpdate() {
        let params = new URLSearchParams(location.search);
        params.set('seed', this.state.seed);
        params.set('populationSize', this.state.populationSize);
        params.set('mean', this.state.mean);
        params.set('stdDev', this.state.stdDev);
        params.set('distType', this.state.distType);
        params.set('sampleSize', this.state.sampleSize);
        params.set('numberOfSamples', this.state.numberOfSamples);
        window.history.replaceState(null, '', '?' + params.toString());
    }
    render() {
        return (
            <>
            <Nav/>
            <div className='container'>
                <h2>Central Limit Theorem</h2>
                <div className='row'>
                    <div className='col-4'>
                        <PopulationForm seed={this.state.seed}
                            populationSize={this.state.populationSize}
                            mean={this.state.mean}
                            stdDev={this.state.stdDev}
                            distType={this.state.distType}
                            sampleSize={this.state.sampleSize}
                            handleGeneratePopulation={
                                this.handleGeneratePopulation}
                            handleChange={this.handleChange}
                            showPopBtn={this.state.populationGraphData ?
                                false : true}/>
                        <SampleForm
                            sampleSize={this.state.sampleSize}
                            numberOfSamples={this.state.numberOfSamples}
                            handleChange={this.handleChange}
                            runSample={this.runSample}
                            sampleMeansIdx={this.state.sampleMeansIdx}
                            handleSampleMeansIdx={
                                this.handleSampleMeansIdx}
                            showSampleBtn={
                                this.state.populationGraphData ?
                                    false : true}/>
                        <SampleRangeSlider
                            numberOfSamples={this.state.numberOfSamples}
                            sampleMeansIdx={this.state.sampleMeansIdx}
                            handleSampleMeansIdx={
                                this.handleSampleMeansIdx}
                            sampleSize={this.state.sampleSize}
                            observationIdx={this.state.observationIdx}
                            observationData={this.state.observationData}
                            handleObservationIdx={this.handleObservationIdx}
                            handleResetSamples={
                                this.handleResetSamples}/>
                    </div>
                    <div className='col-8' style={{maxHeight: '320px'}}>
                        <PopulationGraph
                            populationGraphData={
                                this.state.populationGraphData}
                            samplesGraphData={
                                this.state.samplesGraphData}
                            samplesMax={this.state.samplesMax}
                            observationIdx={this.state.observationIdx}
                            observationData={this.state.observationData}
                            domain={this.state.domain}
                            sampleMean={this.state.sampleMeans ?
                                this.state.sampleMeans[
                                    this.state.sampleMeansIdx] : null}/>
                        <SampleMeansGraph
                            domain={this.state.domain}
                            range={
                                this.state.sampleMeansRange}
                            sampleMeansGraphData={
                                this.state
                                    .sampleMeansGraphData}/>
                    </div>
                </div>
                <hr/>
                <div className="text-center">
                    <button type="button"
                        className="btn btn-danger"
                        id='reset-simulation'
                        onClick={this.handleResetSimulation}>
                        Reset Simulation
                    </button>
                    <CopyToClipboard text={location.href}>
                        <button type="button" className="btn btn-success">
                            Copy Link to Clipboard
                        </button>
                    </CopyToClipboard>
                </div>
            </div>
            </>
        );
    }
}
