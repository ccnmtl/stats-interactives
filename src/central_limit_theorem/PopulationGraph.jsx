import React from 'react';
import PropTypes from 'prop-types';
import {
    VictoryChart, VictoryTheme, VictoryBar,
    VictoryScatter, VictoryAxis} from 'victory';
import * as math from 'mathjs';
math.config({matrix: 'Array'});
import { BAR_FILL, BAR_BORDER, INDICATOR, DOT_STROKE } from '../colors.js';
import { MIN_BIN, MAX_BIN } from './CentralLimitGraph';

export const PopulationGraph  = (
    {populationGraphData, samplesGraphData, samplesMax,
        observationIdx, domain, sampleMean, distType}) => {
    return (
        <>
        <VictoryChart theme={VictoryTheme.material}
            title={'Population and Current Sample'}
            desc={`The population is a ${distType} distribution. The
                distribution of samples largely follows
                that of the population.`}
            padding={{top: 0, left: 10, right: 10, bottom: 25}}
            width={850}
            height={350}
            style={{
                parent: {
                    height: 'inherit',
                    padding: '0.5em 1em 1em 1em',
                }
            }}
            domain={{x: [MIN_BIN, MAX_BIN], y: [0, 1]}}>
            <VictoryAxis
                tickValues={math.range(MIN_BIN, MAX_BIN, true)} />
            {populationGraphData &&
                <VictoryBar data={populationGraphData}
                    alignment='start'
                    barRatio={1}
                    x={0}
                    y={1}
                    style={{ data: { fill: BAR_FILL, stroke: BAR_BORDER,
                        strokeWidth: '2px'} }}/> }
            {samplesGraphData &&
                <VictoryScatter data={samplesGraphData.slice(0, observationIdx)}
                    style={{ data: { fill: 'white', stroke: BAR_BORDER,
                        strokeWidth: DOT_STROKE } }}
                    size={4}
                    x={(datum) => datum[0] + 0.5}
                    y={(datum) => (datum[1] / samplesMax)}/>
            }
            {samplesGraphData &&
                    <VictoryScatter data={
                        samplesGraphData.slice(
                            observationIdx - 1, observationIdx)}
                    style={{ data: { fill: INDICATOR, stroke: BAR_BORDER,
                        strokeWidth: DOT_STROKE } }}
                    size={4}
                    x={(datum) => datum[0] + 0.5}
                    y={(datum) => (datum[1] / samplesMax)}/>
            }
        </VictoryChart>
        </>
    );
};

PopulationGraph.propTypes = {
    populationGraphData: PropTypes.array,
    samplesGraphData: PropTypes.array,
    samplesMax: PropTypes.number,
    observationIdx: PropTypes.number,
    observationData: PropTypes.array,
    domain: PropTypes.array,
    sampleMean: PropTypes.number,
    distType: PropTypes.string,
};

