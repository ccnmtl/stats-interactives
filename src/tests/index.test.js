/* eslint-disable */
import React, {Component} from 'react';
import ReactDOM from 'react-dom';
import {configure, shallow, mount, render} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import {MemoryRouter} from 'react-router-dom';
import {Preview} from '../App.jsx';
import {Nav} from '../Nav.jsx';
import { CentralLimitGraph } from '../central_limit_theorem/CentralLimitGraph';

test('Nav elements render as expected', () => {
    const component = renderer.create(
        <MemoryRouter>
            <Nav />
        </MemoryRouter>,);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
})

test('Preview elements render as expected', () => {
    const component = renderer.create(
        <MemoryRouter>
            <Preview />
        </MemoryRouter>,);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
})
