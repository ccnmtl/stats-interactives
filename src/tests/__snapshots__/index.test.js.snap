// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nav elements render as expected 1`] = `
<nav
  className="navbar navbar-expand-lg navbar-dark"
>
  <a
    className="navbar-brand"
    href="/"
    onClick={[Function]}
  >
    <span
      className="stats"
    >
      Stats
    </span>
    Interactives
    <span
      className="beta"
    >
       Beta
    </span>
  </a>
  <div
    className="navbar-nav-scroll"
  >
    <ul
      className="navbar-nav bd-navbar-nav flex-row"
    >
      <li
        className="nav-item"
      >
        <a
          className="nav-link"
          href="/"
          onClick={[Function]}
        >
          ‚Üê Back to the Collection
        </a>
      </li>
    </ul>
  </div>
  <button
    aria-controls="navbarSupportedContent"
    aria-expanded="false"
    aria-label="Toggle navigation"
    className="navbar-toggler"
    data-target="#navbarSupportedContent"
    data-toggle="collapse"
    type="button"
  >
    <span
      className="navbar-toggler-icon"
    />
     
  </button>
  <div
    className="collapse navbar-collapse"
    id="navbarSupportedContent"
  />
</nav>
`;

exports[`Preview elements render as expected 1`] = `
Array [
  <div
    className="card col-md-6"
  >
    <div
      className="card-body"
    >
      <h2
        className="card-title"
      >
        <a
          href="/central-limit-theorem"
          onClick={[Function]}
        >
          The Central Limit Theorem
        </a>
      </h2>
      <p
        className="card-text"
      >
        In probability theory, the central limit theorem (CLT) establishes that, in some situations, when independent random variables are added, their properly normalized sum tends toward a normal distribution (informally a "bell curve") even if the original variables themselves are not normally distributed.
      </p>
    </div>
  </div>,
  <div
    className="card col-md-6"
  >
    <div
      className="card-body"
    >
      <h2
        className="card-title"
      >
        <a
          href="/linear-regression-model"
          onClick={[Function]}
        >
          Linear Regression Model
        </a>
      </h2>
      <p
        className="card-text"
      >
        A linear regression model assumes that the mean of the response variable is affected by the explanatory variables and that individual observations vary around the mean due to other factors.
      </p>
    </div>
  </div>,
  <div
    className="card col-md-6"
  >
    <div
      className="card-body"
    >
      <h2
        className="card-title"
      >
        <a
          href="/least-squares-estimation"
          onClick={[Function]}
        >
          Least Squares Estimation
        </a>
      </h2>
      <p
        className="card-text"
      >
        Lorem ipsum dolor sit amet, consectetur adipiscing elit. Fusce pharetra lorem blandit, blandit lectus vel, tincidunt leo. Donec sagittis eleifend purus at egestas. Praesent eu mattis est. Etiam et lectus sed libero cursus cursus. Class aptent taciti socios quad litora torquent per conubia nostra, per inceptos himenaeos. Integer a dignissim justo. Suspendisse gravida lectus eget tincidunt.
      </p>
    </div>
  </div>,
  <div
    className="card col-md-6"
  >
    <div
      className="card-body"
    >
      <h2
        className="card-title"
      >
        <a
          href="/sampling-distribution-regression"
          onClick={[Function]}
        >
          Sampling Distribution Regression Coefficients
        </a>
      </h2>
      <p
        className="card-text"
      >
        Lorem ipsum dolor sit amet, consectetur adipiscing elit. Fusce pharetra lorem blandit, blandit lectus vel, tincidunt leo. Donec sagittis eleifend purus at egestas. Praesent eu mattis est. Etiam et lectus sed libero cursus cursus. Class aptent taciti socios quad litora torquent per conubia nostra, per inceptos himenaeos. Integer a dignissim justo. Suspendisse gravida lectus eget tincidunt.
      </p>
    </div>
  </div>,
]
`;
